[Style] Replace FillLayer linked list with strong style typed fixed vector

weinig opened the pull request:
#### 9c502fcf9c71230c41eba0c0b3975431dd4b82fb
<pre>
[Style] Replace FillLayer linked list with strong style typed fixed vector
<a href="https://bugs.webkit.org/show_bug.cgi?id=298113">https://bugs.webkit.org/show_bug.cgi?id=298113</a>

Reviewed by Darin Adler.

Modernizes the background and mask image fill layer code to use
a strong style types and contiguously allocated storage.

To allow for this to work, a new priority bucket, &quot;medium priority&quot;,
was added for CSSProperties.json, for properties that need to be handled
before others, but also can use length values (unlike the &quot;high priority&quot;
bucket). The only two &quot;medium priority&quot; properties for now are `background-image`
and `mask-image`. These two properties also get labeled as &quot;fill-layer-primary&quot;,
meaning that their style building code sets the length of the layers vector
(which is why they need the &quot;medium priority&quot; bucketing).

With the new &quot;fill-layer-primary&quot; concept in place, the style building code
for fill layers was modified so the primary properties create the final vector,
and non-primary properties modify it in place. To reduce unnecessary reliance
on the python for code generation, the implementations of the style building
code were lifted into generic helpers. This matches the pattern already adopted
in the style extractor.

* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/background-repeat-expected.txt:
    - Update results with additional passing subtest results due to allowing all values are of background-repeat at style building time.

* Source/WTF/wtf/RefCountedFixedVector.h:
* Source/WebCore/Headers.cmake:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSToStyleMap.cpp:
* Source/WebCore/css/CSSToStyleMap.h:
* Source/WebCore/css/CSSValueList.h:
* Source/WebCore/css/scripts/process-css-properties.py:
* Source/WebCore/css/scripts/test/TestCSSProperties.json:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSPropertyNames.gperf:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSPropertyNames.h:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSPropertyParsing.cpp:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSStyleProperties+PropertyNames.idl:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StyleBuilderGenerated.cpp:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StyleExtractorGenerated.cpp:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StyleInterpolationWrapperMap.cpp:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StylePropertyShorthandFunctions.cpp:
* Source/WebCore/css/values/CSSValueAggregates.h:
* Source/WebCore/css/values/CSSValueTypes.h:
* Source/WebCore/html/HTMLInputElement.cpp:
* Source/WebCore/page/ElementTargetingController.cpp:
* Source/WebCore/page/InteractionRegion.cpp:
* Source/WebCore/page/LocalFrameView.cpp:
* Source/WebCore/rendering/BackgroundPainter.cpp:
* Source/WebCore/rendering/BackgroundPainter.h:
* Source/WebCore/rendering/InlineBoxPainter.cpp:
* Source/WebCore/rendering/InlineBoxPainter.h:
* Source/WebCore/rendering/ReferencedSVGResources.cpp:
* Source/WebCore/rendering/RenderBox.cpp:
* Source/WebCore/rendering/RenderBox.h:
* Source/WebCore/rendering/RenderBoxModelObject.h:
* Source/WebCore/rendering/RenderElement.cpp:
* Source/WebCore/rendering/RenderElement.h:
* Source/WebCore/rendering/RenderImage.cpp:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
* Source/WebCore/rendering/RenderLayerModelObject.cpp:
* Source/WebCore/rendering/RenderTableCell.cpp:
* Source/WebCore/rendering/RenderTheme.h:
* Source/WebCore/rendering/RenderView.cpp:
* Source/WebCore/rendering/TextPaintStyle.cpp:
* Source/WebCore/rendering/style/FillLayer.cpp: Removed.
* Source/WebCore/rendering/style/FillLayer.h: Removed.
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleConstants.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/StyleBackgroundData.cpp:
* Source/WebCore/rendering/style/StyleBackgroundData.h:
* Source/WebCore/rendering/style/StyleMiscNonInheritedData.cpp:
* Source/WebCore/rendering/style/StyleMiscNonInheritedData.h:
* Source/WebCore/rendering/svg/RenderSVGTextPath.h:
* Source/WebCore/rendering/svg/SVGRenderTreeAsText.cpp:
* Source/WebCore/rendering/svg/legacy/SVGResources.cpp:
* Source/WebCore/rendering/svg/legacy/SVGResourcesCache.cpp:
* Source/WebCore/style/StyleAdjuster.cpp:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorCustom.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
* Source/WebCore/style/StyleInterpolationWrappers.h:
* Source/WebCore/style/StylePendingResources.cpp:
* Source/WebCore/style/values/StyleValueTypes.h:
* Source/WebCore/style/values/backgrounds/StyleBackgroundLayer.cpp: Added.
* Source/WebCore/style/values/backgrounds/StyleBackgroundLayer.h: Added.
* Source/WebCore/style/values/backgrounds/StyleBackgroundSize.cpp: Added.
* Source/WebCore/style/values/backgrounds/StyleBackgroundSize.h: Added.
* Source/WebCore/style/values/backgrounds/StyleFillLayers.h: Added.
* Source/WebCore/style/values/backgrounds/StyleRepeatStyle.cpp: Added.
* Source/WebCore/style/values/backgrounds/StyleRepeatStyle.h: Added.
* Source/WebCore/style/values/masking/StyleMaskLayer.cpp: Added.
* Source/WebCore/style/values/masking/StyleMaskLayer.h: Added.
* Source/WebCore/style/values/motion/StyleOffsetPosition.cpp:
* Source/WebCore/style/values/text-decoration/StyleTextDecorationLine.cpp:
* Tools/Scripts/webkitpy/style/checkers/jsonchecker.py:

Canonical link: <a href="https://commits.webkit.org/299407@main">https://commits.webkit.org/299407@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/50b0f6e9830dd35d33918ed39b1c47a9b5fdc882

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/118923 "15 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/38604 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/29255 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/125123 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/70986 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/120801 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/39300 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/47186 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/90252 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/59768 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/121876 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/31330 "Build is in progress. Recent messages:OS: Sequoia (15.5), Xcode: 16.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running run-layout-tests-in-stress-mode; 8 new passes 3 flakes") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/106622 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/70757 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/118284 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/30391 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/24736 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/68776 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/111041 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/100774 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/24924 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/128163 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/117437 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/45830 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/34623 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/98916 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/46196 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/102842 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/98695 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25082 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/44149 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/22149 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/42395 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/45700 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/51378 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/146135 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/45166 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/37593 "Found 11 new JSC stress test failures: wasm.yaml/wasm/lowExecutableMemory/executable-memory-oom.js.default-wasm, wasm.yaml/wasm/stress/cc-int-to-int-cross-module-with-exception.js.default-wasm, wasm.yaml/wasm/stress/cc-int-to-int-jit-to-llint.js.default-wasm, wasm.yaml/wasm/stress/wasm-js-string-builtins.js.default-wasm, wasm.yaml/wasm/stress/wasm-js-string-builtins.js.wasm-bbq, wasm.yaml/wasm/stress/wasm-js-string-builtins.js.wasm-bbq-no-consts, wasm.yaml/wasm/stress/wasm-js-string-builtins.js.wasm-collect-continuously, wasm.yaml/wasm/stress/wasm-js-string-builtins.js.wasm-eager, wasm.yaml/wasm/stress/wasm-js-string-builtins.js.wasm-eager-jettison, wasm.yaml/wasm/stress/wasm-js-string-builtins.js.wasm-no-cjit ... (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/48510 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/46850 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/CSSValueList.h:
```diff
@@ -42,17 +42,23 @@ class CSSValueContainingVector : public CSSValue {
         constexpr bool operator==(const iterator& other) const { return index == other.index; }
 
-        const CSSValueContainingVector& vector;
+        const CSSValueContainingVector* vector;
```
What's the motivation for this change?

weinig replied:
Oh! I meant to add a static assert for that. It’s to make std::range compatible.

darinadler commented on Source/WebCore/rendering/BackgroundPainter.cpp:
```diff
@@ -139,23 +148,26 @@ void BackgroundPainter::paintFillLayers(const Color& color, const FillLayer& fil
 
     if (shouldDrawBackgroundInSeparateBuffer) {
-        paintFillLayer(color, *layers.last(), rect, bleedAvoidance, { }, { }, op, backgroundObject, BaseBackgroundColorOnly);
+        paintFillLayerImpl(color, fillLayers.last(), true, rect, bleedAvoidance, { }, { }, op, backgroundObject, BaseBackgroundColorOnly);
```
Unclear that the `true` here is "last layer". No obvious simple solution comes to mind. Named boolean constant is kinda weak. Enumeration is a bit heavy handed.

Maybe the function should assume it's the last layer by default and there could be an optional "not the last layer" argument.

weinig replied:
Or maybe a struct with two parameters:

```c++
struct LayerToPaint {
    Layer& layer;
    bool isLast;
};
```

so then this could be

```c++
paintFillLayerImpl(color, { .layer = fillLayers.last(), .isLast = true }, rect, bleedAvoidance, { }, { }, op, backgroundObject, BaseBackgroundColorOnly);
```

darinadler commented on Source/WebCore/rendering/BackgroundPainter.h:
```diff
@@ -71,20 +71,27 @@ class BackgroundPainter {
-    static LayoutSize calculateFillTileSize(const RenderBoxModelObject&, const FillLayer&, const LayoutSize& positioningAreaSize);
+    template<typename LayerList> void paintFillLayersImpl(const Color&, const LayerList&, const LayoutRect&, BleedAvoidance, CompositeOperator, RenderElement* backgroundObject = nullptr) const;
+    template<typename Layer> void paintFillLayerImpl(const Color&, const Layer&, bool isLastLayer, const LayoutRect&, BleedAvoidance, const InlineIterator::InlineBoxIterator&, const LayoutRect& backgroundImageStrip = { }, CompositeOperator = CompositeOperator::SourceOver, RenderElement* backgroundObject = nullptr, BaseBackgroundColorUsage = BaseBackgroundColorUse) const;
+    template<typename Layer> static BackgroundImageGeometry calculateFillLayerImageGeometryImpl(const RenderBoxModelObject&, const RenderLayerModelObject* paintContainer, const Layer&, const LayoutPoint& paintOffset, const LayoutRect& borderBoxRect, std::optional<FillBox> overrideOrigin = std::nullopt);
```
I'm not a huge fan of the word _impl_. I wonder if there's another naming style that would work here. No suggestion in mind though.

weinig replied:
Sometimes I use `Generic` as the suffix I think, not sure I like that any more (or less).

darinadler commented on Source/WebCore/rendering/InlineBoxPainter.h:
```diff
@@ -55,8 +54,8 @@ class InlineBoxPainter {
-    void paintFillLayers(const Color&, const FillLayer&, const LayoutRect& paintRect, CompositeOperator);
-    void paintFillLayer(const Color&, const FillLayer&, const LayoutRect& paintRect, CompositeOperator);
+    template<typename Layers> void paintFillLayers(const Color&, const Layers&, const LayoutRect& paintRect, CompositeOperator);
+    template<typename Layer> void paintFillLayer(const Color&, const Layer&, bool isLastLayer, const LayoutRect& paintRect, CompositeOperator);
```
How did you decide on `template` vs. `auto` here?

weinig replied:
My general rule is if the typename is needed I use `template`, but it looks like I didn't apply that here. Should probably be `auto`.

darinadler commented on Source/WebCore/style/values/backgrounds/StyleBackgroundLayer.cpp:
```diff
@@ -0,0 +1,101 @@
+        && m_origin == o.m_origin
+        && m_repeat == o.m_repeat
+        && m_size == o.m_size;
+}
```
Can we use the default instead?

weinig replied:
No, we don’t want to include m_clipMax. 

Will add a comment to make that reasoning clear.

weinig committed:
[Style] Replace FillLayer linked list with strong style typed fixed vector
https://bugs.webkit.org/show_bug.cgi?id=298113

Reviewed by Darin Adler.

Modernizes the background and mask image fill layer code to use
a strong style types and contiguously allocated storage.

To allow for this to work, a new priority bucket, "medium priority",
was added for CSSProperties.json, for properties that need to be handled
before others, but also can use length values (unlike the "high priority"
bucket). The only two "medium priority" properties for now are `background-image`
and `mask-image`. These two properties also get labeled as "fill-layer-primary",
meaning that their style building code sets the length of the layers vector
(which is why they need the "medium priority" bucketing).

With the new "fill-layer-primary" concept in place, the style building code
for fill layers was modified so the primary properties create the final vector,
and non-primary properties modify it in place. To reduce unnecessary reliance
on the python for code generation, the implementations of the style building
code were lifted into generic helpers. This matches the pattern already adopted
in the style extractor.

* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/background-repeat-expected.txt:
    - Update results with additional passing subtest results due to allowing all values are of background-repeat at style building time.

* Source/WTF/wtf/RefCountedFixedVector.h:
* Source/WebCore/Headers.cmake:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSToStyleMap.cpp:
* Source/WebCore/css/CSSToStyleMap.h:
* Source/WebCore/css/CSSValueList.h:
* Source/WebCore/css/scripts/process-css-properties.py:
* Source/WebCore/css/scripts/test/TestCSSProperties.json:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSPropertyNames.gperf:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSPropertyNames.h:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSPropertyParsing.cpp:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSStyleProperties+PropertyNames.idl:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StyleBuilderGenerated.cpp:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StyleExtractorGenerated.cpp:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StyleInterpolationWrapperMap.cpp:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StylePropertyShorthandFunctions.cpp:
* Source/WebCore/css/values/CSSValueAggregates.h:
* Source/WebCore/css/values/CSSValueTypes.h:
* Source/WebCore/html/HTMLInputElement.cpp:
* Source/WebCore/page/ElementTargetingController.cpp:
* Source/WebCore/page/InteractionRegion.cpp:
* Source/WebCore/page/LocalFrameView.cpp:
* Source/WebCore/rendering/BackgroundPainter.cpp:
* Source/WebCore/rendering/BackgroundPainter.h:
* Source/WebCore/rendering/InlineBoxPainter.cpp:
* Source/WebCore/rendering/InlineBoxPainter.h:
* Source/WebCore/rendering/ReferencedSVGResources.cpp:
* Source/WebCore/rendering/RenderBox.cpp:
* Source/WebCore/rendering/RenderBox.h:
* Source/WebCore/rendering/RenderBoxModelObject.h:
* Source/WebCore/rendering/RenderElement.cpp:
* Source/WebCore/rendering/RenderElement.h:
* Source/WebCore/rendering/RenderImage.cpp:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
* Source/WebCore/rendering/RenderLayerModelObject.cpp:
* Source/WebCore/rendering/RenderTableCell.cpp:
* Source/WebCore/rendering/RenderTheme.h:
* Source/WebCore/rendering/RenderView.cpp:
* Source/WebCore/rendering/TextPaintStyle.cpp:
* Source/WebCore/rendering/style/FillLayer.cpp: Removed.
* Source/WebCore/rendering/style/FillLayer.h: Removed.
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleConstants.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/StyleBackgroundData.cpp:
* Source/WebCore/rendering/style/StyleBackgroundData.h:
* Source/WebCore/rendering/style/StyleMiscNonInheritedData.cpp:
* Source/WebCore/rendering/style/StyleMiscNonInheritedData.h:
* Source/WebCore/rendering/svg/RenderSVGTextPath.h:
* Source/WebCore/rendering/svg/SVGRenderTreeAsText.cpp:
* Source/WebCore/rendering/svg/legacy/SVGResources.cpp:
* Source/WebCore/rendering/svg/legacy/SVGResourcesCache.cpp:
* Source/WebCore/style/StyleAdjuster.cpp:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorCustom.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
* Source/WebCore/style/StyleInterpolationWrappers.h:
* Source/WebCore/style/StylePendingResources.cpp:
* Source/WebCore/style/values/StyleValueTypes.h:
* Source/WebCore/style/values/backgrounds/StyleBackgroundLayer.cpp: Added.
* Source/WebCore/style/values/backgrounds/StyleBackgroundLayer.h: Added.
* Source/WebCore/style/values/backgrounds/StyleBackgroundSize.cpp: Added.
* Source/WebCore/style/values/backgrounds/StyleBackgroundSize.h: Added.
* Source/WebCore/style/values/backgrounds/StyleFillLayers.h: Added.
* Source/WebCore/style/values/backgrounds/StyleRepeatStyle.cpp: Added.
* Source/WebCore/style/values/backgrounds/StyleRepeatStyle.h: Added.
* Source/WebCore/style/values/masking/StyleMaskLayer.cpp: Added.
* Source/WebCore/style/values/masking/StyleMaskLayer.h: Added.
* Source/WebCore/style/values/motion/StyleOffsetPosition.cpp:
* Source/WebCore/style/values/text-decoration/StyleTextDecorationLine.cpp:
* Tools/Scripts/webkitpy/style/checkers/jsonchecker.py:

Canonical link: https://commits.webkit.org/299407@main

ddkilzer commented on Source/WebCore/rendering/style/FillLayer.h:
```diff
@@ -1,229 +0,0 @@
-    LengthSize m_sizeLength;
-    FillRepeatXY m_repeat;
-
-    static constexpr unsigned FillBoxBitWidth = 3;
-
-    PREFERRED_TYPE(FillAttachment) unsigned m_attachment : 2;
-    PREFERRED_TYPE(FillBox) unsigned m_clip : FillBoxBitWidth;
-    PREFERRED_TYPE(FillBox) unsigned m_origin : FillBoxBitWidth;
```
This change regressed this fix made to Source/WebCore/rendering/style/FillLayer.h:

> Fix -Wpreferred-type-bitfield-enum-conversion warnings in WebCore
> <https://bugs.webkit.org/show_bug.cgi?id=297439>
> <rdar://158245532>
> <https://commits.webkit.org/298737@main>

(It's only caught by newer versions of clang, which is why builds didn't fail.)

ddkilzer commented on Source/WebCore/style/values/backgrounds/StyleBackgroundLayer.h:
```diff
@@ -0,0 +1,123 @@
+    BackgroundSize m_size;
+    RepeatStyle m_repeat;
+
+    PREFERRED_TYPE(FillAttachment) unsigned m_attachment : 2;
+    PREFERRED_TYPE(FillBox) unsigned m_clip : 3;
+    PREFERRED_TYPE(FillBox) unsigned m_origin : 2;
+    PREFERRED_TYPE(BlendMode) unsigned m_blendMode : 5;
+
+    PREFERRED_TYPE(FillBox) mutable unsigned m_clipMax : 2; // maximum m_clip value from this to bottom layer
```
This change regressed this fix made to Source/WebCore/rendering/style/FillLayer.h:

> Fix -Wpreferred-type-bitfield-enum-conversion warnings in WebCore
> <https://bugs.webkit.org/show_bug.cgi?id=297439>
> <rdar://158245532>
> <https://commits.webkit.org/298737@main>

(It's only caught by newer versions of clang, which is why builds didn't fail.)
